babelfish-for-postgresql/babelfish_extensions
# Support TSQL ownership chaining within same DB in procedure

### Description

This commit supports forming ownership chain while executing the procedure such that when referenced object(in any particular statement) and procedure has same owner then invoker will be able to execute that statement even if it doesn't have explicit permissions on referenced object.
To implement above functionality we've implemented pg_*_aclcheck hooks which
- check whether current statement execution is inside procedure or not by looking at the PLtsql_exec_state stack.
- If it is inside procedure then it compares the referenced object and procedure's owner.
- If both are same then this hook bypasses the permission check and allows execution of that statement.

Task: BABEL-3014
Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

### Test Scenarios Covered ###
* **Use case based -**
    * Object calls inside function/procedure
        * Covered table, view, procedure, function
    * Nested Scenarios
    * Chained object references
        * Procedure -> View -> View -> Table
    * Check ownership chaining after granting permissions on partial columns of a table
* **Boundary conditions -** NA
* **Arbitrary inputs -** NA
* **Negative test cases -**
    * Procedure/Function execution after using objects with different owner
    * Call/Exec using sp_prepare - Dynamic SQL
    * DDL/Truncate Table
* **Minor version upgrade tests -** NA
* **Major version upgrade tests -** NA
* **Performance tests -** NA
* **Tooling impact -** NA
* **Client tests -** NA

### Check List

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
